hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -29,77 +29,342 @@
 import com.hazelcast.query.impl.predicates.OrPredicate;
 import com.hazelcast.query.impl.predicates.RegexPredicate;
 
+import java.util.Date;
+
 /**
  * A utility class to create {@link com.hazelcast.query.Predicate} instances.
+ * <p>
+ * <b>Special Attributes</b>
+ * <p>
+ * The predicate creation methods accepting an attribute name support two special attributes:
+ * <ul>
+ * <li>{@code "__key"} - instructs the predicate to act on a key associated with an item.
+ * <li>{@code "this"} - instructs the predicate to act on a value associated with an item.
+ * </ul>
+ * <p>
+ * <b>Handling of {@code null}</b>
+ * <p>
+ * The predicate creation methods can accept {@code null} as a value to compare with or a pattern to match against
+ * if and only if that is explicitly stated in the method documentation. In this case, the usual {@code null} equality
+ * logic applies: if {@code null} is provided, the predicate passes an item if and only if the value stored under the
+ * item attribute in question is also {@code null}.
+ * <p>
+ * Special care must be taken while comparing with {@code null} values <i>stored</i> inside items being filtered
+ * through the predicates created by the following methods: {@link #greaterThan(String, Comparable) greaterThan},
+ * {@link #greaterEqual(String, Comparable) greaterEqual}, {@link #lessThan(String, Comparable) lessThan},
+ * {@link #lessEqual(String, Comparable) lessEqual}, {@link #between(String, Comparable, Comparable) between}. The
+ * predicates produced by these methods intentionally violate {@link Comparable} contract by not throwing
+ * {@link NullPointerException} for {@code null} values. Instead, they always evaluate to {@code false} and therefore
+ * never pass such items.
+ * <p>
+ * <b>Type Conversion</b>
+ * <p>
+ * If the type of the stored value doesn't match the type of the value provided to the predicate, type conversion is
+ * performed before predicate evaluation. The provided value is converted to match the type of the stored attribute value.
+ * If there is no conversion exists to match the type, {@link IllegalArgumentException} is thrown.
```

Comment:
"If no conversion matching the type exists"

Reply:
fixed